Avoid leaking unprotected SSH private key files onto rescue medium

OliverO2 opened the pull request:
**Note:** This is not about logging into the rescue system via SSH (incoming access). It is about accessing _other systems_ from a running rescue system via SSH private keys stored on the rescue medium (outgoing access).

#### Problem

The 'root' account on the original system may contain SSH private key files without passphrase protection. Such unprotected private key files are typically used to access other systems during unattended operations (e.g. automatic backup to a network server). This kind of access can be useful during restore, too.

Security then depends entirely on keeping such private keys secret, which might be compromised if unprotected private key files leak onto an unencrypted rescue medium. Anyone with physical access to the rescue medium would be allowed unauthenticated access to each account on the network authorizing one of those private keys.

#### Solution

Avoid leaking SSH private key files onto the rescue medium in unprotected form:

This PR enables ReaR to copy the SSH client configuration for 'root' with the option of passphrase-protecting otherwise unprotected private keys. The user can choose between
- copying previously unprotected private keys with added passphrase-protection,
- copying previously unprotected private keys as-is (with a proper warning),
- not copying unprotected private keys at all.

Private key files which already contain a passphrase are assumed to be sufficiently secure and are copied as-is.

The PR also aims to optimize usability during 'rear recover' by asking the user just once for any required SSH passphrase (using ssh-agent).

Tested on Ubuntu 16.04.3 LTS.

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -54,5 +57,110 @@ if has_binary sshd; then
+
+                        while true; do
+                            LogUserOutput "SSH client passphrase (or empty, or 'UnProtected'): \c"
+                            read -s passphrase <&6
```
Can't you use the new `UserInput` function for this? That way you could also handle setting an answer for it.

OliverO2 replied:
I did first, but then `UserInput` would copy the passphrase into the log file as plain text, which is a no-go. So doing so would require a careful change to `UserInput`.

schlomo replied:
Yes, go ahead and improve `UserInput` if you need it.

jsmeix replied:
I will enhance the UserInput function to be silent
if requested (i.e. via an additional option).

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -54,5 +57,110 @@ if has_binary sshd; then
+
+                                *)
+                                    LogUserOutput "Repeat SSH client passphrase: \c"
+                                    read -s repeated_passphrase <&6
```
As above, I wouldn't ask for a password repeat. ReaR is meant to be run in an automated fashion and not manually. I would even go so far as to wonder why we should interactively ask for this password here. Better we should support a variable for the password. If a user really wants to enter this password for *every* `rear mkrescue` run then the user should set the variable via `read` or `UserInput` in the `local.conf`.

OliverO2 replied:
The primary objective is not to leak any secret into the rescue system. If the passphrase is supplied via a configuration variable in `local.conf` or `site.conf`, it would end up in the rescue system:
```
# ll etc/rear
total 16K
-rw-r--r-- 1 root root 476 Sep 18 23:37 local.conf
-rw-r--r-- 1 root root 275 Sep 18 23:37 os.conf
-rw-r--r-- 1 root root 888 Sep 18 23:37 rescue.conf
-rw-r--r-- 1 root root 137 Sep 18 23:37 site.conf
```

schlomo replied:
Sorry, @jsmeix worked hard to remove all the blocking reads in ReaR. If `UserInput` doesn't yet know how to handle passwords then we should implement this now. IMHO it would be even better to get away without the need to prompt for passwords.

Setting the variable in the configuration doesn't mean that is has to be hard coded. It can be read there with Bash or any other method. If the user puts this configuration into `mkrescue.conf` then it will only be active for the`mkrescue` flow.

Separating between using the passphrase (in these scripts) and providing it (in the configuration) also allows for much greater flexibility. Maybe some user has a secure storage for the passphrase that is not part of the backup. Or maybe he triggers the `rear mkrescue` call from somewhere else and provides the password on the command line or as an environment variable. Lots lots of options open up if we separate those concerns.

OliverO2 replied:
> Setting the variable in the configuration doesn't mean that is has to be hard coded. It can be read there with Bash or any other method.

This is exactly what I was thinking when I woke up this morning. ;-)

> If the user puts this configuration into 'mkrescue.conf' then it will only be active for the `mkrescue` flow.

I couldn't find any documentation on this. And I'd still prefer to separate secrets from configuration.

> Lots lots of options open up if we separate those concerns.

From my point of view, lots of options are a problem in security, as they make it easy to mess things up with homegrown, poorly understood and poorly tested setups. I'd rather prefer a single, well-understood option which is just "good enough" and works safely even for non-experts.

schlomo replied:
I know that for many ReaR users it takes a bit to get used to the idea that ReaR **configuration** is actually regular **Bash code** that allows them to do what they want.

ReaR has probably thousands of undocumented features :sob: I just happen to remember a lot of the stuff I did 10 years ago.
```bash
~/src/rear/rear$ git grep WORKFLOW.conf
usr/sbin/rear:test -r "$CONFIG_DIR/$WORKFLOW.conf" && Source "$CONFIG_DIR/$WORKFLOW.conf" || true
usr/share/rear/lib/config-functions.sh:    # This is needed, so that they can be overridden in $WORKFLOW.conf
```

If you want to keep using `local.conf` then you can check the `$WORKFLOW` variable yourself:
```bash
if [[ "$WORKFLOW" == mk* ]] ; then
  read -r -s -p "Enter passphrase: " 
  SSH_UNPROTECTED_KEYS_SET_PASSPHRASE="$REPLY" # support passwords with $IFS in them
  read -r -s -p "Confirm passphrase: "
  [ "$REPLY" = "$SSH_UNPROTECTED_KEYS_SET_PASSPHRASE" ] || Error "Passphrase confirmation doesn't match"
fi
```

What do you think about simplifying stuff by removing the SSH key modifications and making do with the improvement of not copying unprotected SSH keys unless specifically configured to do so?

jsmeix replied:
@schlomo
only for the fun of it because I just happen to remember
a lot of the stuff I did 10 days or weeks ago ;-)

Plain
<pre>
read -r -s -p "Enter passphrase: "
</pre>
never did work because STDERR is redirected into the log
and 'read' shows its prompt on STDERR.

Nowadays also STDOUT is redirected into the log
so that one should use
<pre>
read -r -s -p "Enter passphrase: " 0<&6 1>&7 2>&8
</pre>
cf. "What to do with stdout and stderr" at
https://github.com/rear/rear/wiki/Coding-Style
The '0<&6' is currently not required but I prefer to use
the whole '0<&6 1>&7 2>&8' term to make it clear that
"all I/O via the original FDs" is meant.

schlomo replied:
Yes, you are right @jsmeix. Can we get a convenience function for that, something like this?
```bash
function without_redirects {
  eval "$@" 0<&6 1>&7 2>&8
}

without_redirects read -r -s -p "Prompt me: "
```

jsmeix replied:
Yes we can!
But not today ;-)

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -54,5 +57,110 @@ if has_binary sshd; then
+                    case "$passphrase" in
+                        "")
+                            Log "Ignoring unprotected private key $original_file"
+                            rm "$target_file"
```
First copy and then remove? I'd really prefer forward logic, if possible. Like copy only if it is protected.

OliverO2 replied:
I did so first. But then we'd change the original private key file by just testing for password-protection. Trying to assign an empty passphrase for testing changes not only the private key file's mtime, but also its contents.

I'd rather leave the original private key file untouched and this seemed the simpler solution. An alternative would be to copy twice, first to some staging area and then into the target directory. Doing so would require more cleanup.

schlomo replied:
Yes, `ssh-keygen` must modify the file. 

I think your entire code will become easier to read and understand if you split the `$passphrase` variable into two proper ReaR variables that are also explained and set in `default.conf`:

1. SSH_KEEP_UNPROTECTED_KEYS=false is the default and can be enabled to allow unprotected keys
1. SSH_UNPROTECTED_KEYS_SET_PASSPHRASE=foobar can be set to protect the keys in the rescue media

Such variables will enable you to write simple if statements that read much better and don't overload the passhphrase variable with multiple meanings.

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -54,5 +57,110 @@ if has_binary sshd; then
+
+                        *)
+                            Log "Copying private key $original_file, adding passphrase-protection"
+                            if ! ssh-keygen -p -P '' -N "$passphrase" -f "$target_file"; then
```
If you  change the case to a nested if then you can copy the file here. Pseudocode like this:
```bash
if test "$passphrase" ; then
  cp $v -p "$original_file" "$target_dir" 1>&2 || Error "Could not copy $original_file to $target_dir"
  if [ "$passphrase" != UnProtected ] ; then
    ssh-keygen -p -P '' -N "$passphrase" -f "$target_file" || Error "Could not set passphrase on SSH key $target_file"
  fi
fi
```

IMHO this reads much shorter and more straight-forward.

OliverO2 replied:
W.r.t. copying, see above.

I don't mind the short form for the error case, but I'd rather keep a log entry for each private key handled.

schlomo replied:
Log entry is fine, I wouldn't bother the user in the verbose output for stuff that "just works".

schlomo commented on usr/share/rear/setup/default/005_ssh_agent_start.sh:
```diff
@@ -0,0 +1,11 @@
+if has_binary ssh-agent; then
+    Log "Starting up ssh-agent"
+    eval "$(ssh-agent -s)"
+    echo -e "\nHost *\nAddKeysToAgent yes\n" >> /root/.ssh/config
```
According to https://www.openssh.com/txt/release-7.2 this option was added in openssh 7.2 in February 2016. Are you sure that all supported distros have this or a newer version? What happens with older ssh? Why not call `ssh-add` to add all the keys?

(Thanks for pointing me to this option, I didn't know about it)

Another thought: Should we indicate to the user that the SSH key passphrase was changed or set?

OliverO2 replied:
1. Actually I'm not aware about ssh versions on the other distros. Hopefully others could point out some details.
1. An older ssh might just ignore the option, but I don't know für sure. Should probably check this.
1. I was thinking about using `ssh-add` first but wanted to ask the user for a password only whenever it is actually used for an ssh call. After all, we cannot not know a priori which ssh invocations require which private key, if at all.
1. Regarding indications on setting/changing passphrases: Currently, the passphrase question will come up only if required. So the user knows that she would have to remember the passphrase just entered. The user will also be informed by `LogUserOutput "NOTE: ...` if another previously protected private key (with a potentially different passphrase) exists. The user will then find complete details in the log. Anything missing here?

schlomo replied:
First of all, I misunderstood your code. You are right, the agent only runs in the context of a ReaR flow. The "right" way to kill the agent is with the help of `AddExitTask` which you set right after successfully starting the agent.

I imagine that Ubuntu 14, CentOS 5 or SLES 11 will be a good test case for older SSH, please check that this solution also works with older SSH.

I like your goal of only asking passphrases for keys that are actually used. We might have to accept the fact that this goal only works on newer systems. For older systems we still have to provide a useful user experience.

"just entered" only applies to us ReaR developers. Most users employ ReaR as an "install & forget" solution where they automate the mkrescue and don't remember much from the ReaR setup when they actually use the recovery media. That's why I think some guidance will be helpful here.

schlomo commented on usr/share/rear/wrapup/default/970_ssh_agent_stop.sh:
```diff
@@ -0,0 +1,9 @@
+# This file is part of Relax-and-Recover, licensed under the GNU General
+# Public License. Refer to the included COPYING for full text of license.
+
+if [ -n "$SSH_AGENT_PID" ]; then
```
Why? If the user aborts the recovery and tries again it won't work.

Better leave it running till the user reboots the system. I consider the entire lifetime of the rescue system as a single user session for root.

OliverO2 replied:
Hmm, we are in the `rear recover` workflow here, so `ssh` invocations from the login shell would not be aware of `ssh-agent` anyway. We simply cannot export the environment variables to the enclosing shell.

Re-trying an aborted recovery would set up another ssh-agent, wouldn't it?

OliverO2 replied:
Thanks a lot for your thorough review! I'll wait for additional feedback before I come up with commits addressing outstanding issues.

schlomo replied:
Please replace this script file with an `AddExitTask "ssh-agent -k >/dev/null"` or something like this in the `005_ssh_agent_start.sh` script. That way the agent will be killed also when ReaR aborts with an error.

Alternatively, run the agent as a system-level service that starts at boot time and writes the variables to one of the files read during login.

The benefit of that would be that 

1. you don't need to care about agent start/stop within ReaR
1. the user benefits from the agent also when he works manually with ssh outside of ReaR
1. repeated ReaR runs don't ask for the same passphrase (which was your original goal)

OliverO2 replied:
Currently looking at this.

> repeated ReaR runs don't ask for the same passphrase (which was your original goal)

Actually, it wasn't about repeated ReaR runs but repeated ssh invocations during one run of `rear recover`. Over here, the backup service is accessed first to
1. list available backups and
2. to perform the actual restore.

Anyway, it would be nice to have `ssh-agent` support during the whole login session. But it shouldn't complicate things.

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -54,5 +57,110 @@ if has_binary sshd; then
+                # Private key files will be copied, examined and, if desired, passphrase-protected
+
+                target_file="$target_dir/$(basename "$original_file")"
+                cp -p "$original_file" "$target_file"
```
In both cases you copy the file. In the case of a key file you check it an optionally remove it.

I think it would be even simpler to first copy the entire `/root/.ssh` directory as it is and then iterate over the private key files and delete those that are not compliant. Less code, less `case` and less complexity.

OliverO2 replied:
The idea was to copy well-known files only. If the admin leaves other security-related files in this well-protected directory, we'd still be safe. Same goes with new key file types which might come up in the future. Better not copy anything that we don't really know about to an insecure rescue medium.

schlomo replied:
I would copy all of the SSH config directory and fix problems when they arise. Typically there is also some delay between new features showing up and between those features being used.

In any case, you could collapse that all into a single `cp` command with multiple source arguments that represent your whitelist of known files to copy.

OliverO2 replied:
Considered that, too. I'd still end up with a case to deal with private key files matching `id_*` but not `id_*.pub`. See my upcoming commit if you'd just consider it good enough.

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -1,50 +1,53 @@
 # Public License. Refer to the included COPYING for full text of license.
 
-if has_binary sshd; then
+if has_binary ssh || has_binary sshd; then
```
You can do `has_binary ssh || has_binary sshd || return` to bail out early to skip SSH treatment. That will save you another `if` and make the script easier to read.

OliverO2 replied:
True. This structure was just inherited. It's OK for me to change this.

jsmeix commented:
Via https://github.com/rear/rear/pull/1501
there is a new confidential mode (via '-C')
for the UserInput function that does not log
possibly confidentia data.

jsmeix commented:
I did not follow all the details here only a side note
regarding the comment from @schlomo 
https://github.com/rear/rear/pull/1500#discussion_r139680488
<pre>
not copying unprotected SSH keys unless
specifically configured to do so
</pre>

@OliverO2 
in general no confidential stuff should be copied
into the ReaR recovery system unless the user
specifically configured ReaR to do so, cf.
https://github.com/rear/rear/pull/1472#issuecomment-328459748

schlomo commented on usr/share/rear/conf/default.conf:
```diff
@@ -805,6 +805,20 @@ CLONE_ALL_USERS_GROUPS="no"
+# E.g., create
+#   echo "my_recover_passphrase" > /root/.ssh/rear-recover-passphrase && chmod u=rw,go=- /root/.ssh/rear-recover-passphrase
+# then use
+#   SSH_PRIVATE_KEYS_RECOVER_PASSPHRASE="$(cat /root/.ssh/rear-recover-passphrase 2>/dev/null)"
```
`$(< file)` is more concise and avoids the external program call. If you want to use cat then please don't toss the error output so that it can show up in the log.

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -3,164 +3,97 @@
+# Set the SSH root password; if pw is hashed just copy it otherwise use openssl (for backward compatibility)
+if [[ "$SSH_ROOT_PASSWORD" ]] ; then
+    case "$SSH_ROOT_PASSWORD" in
+    '$1$'*) echo "root:$SSH_ROOT_PASSWORD:::::::" > $ROOTFS_DIR/etc/shadow ;;
```
Please update from master first in case #1503 gets merged before yours.

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -3,164 +3,97 @@
-            read subsys sftp original_file junk < <( grep sftp /etc/sshd_co[n]fig /etc/ssh/sshd_co[n]fig /etc/openssh/sshd_co[n]fig /etc/centrifydc/ssh/sshd_co[n]fig 2>/dev/null )
-            echo $original_file
-        )
+COPY_AS_IS=( "${COPY_AS_IS[@]}" /etc/ssh* /root/.ssh/[a]uthorized_keys /root/.shos[t]s )
```
Please look at https://linux.die.net/man/1/ssh under Files. There are many more files in the `/root/.ssh` than those that you copy.

I don't agree with this change in ReaR behavior. So far we copy all files. IMHO this should stay.

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -3,164 +3,97 @@
+# copy ssh user
+# getent will return all entries that match the key(s) exactly - most systems use 'sshd', some may use 'ssh', none should use both.
+# Only the first line (first returned entry) will be used by 'read' in 'IFS=: read ... <<<"$PASSWD_SSH"', so we ask for sshd first.
+PASSWD_SSH=$(getent passwd sshd ssh)
```
I still fail to see why we don't use `CLONE_USERS+( ssh sshd)` here instead. Then the only thing missing is the home directory, which should probably done in `build/default/500_patch_sshd_config.sh`

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -3,164 +3,97 @@
+    cp -p "$original_file" "$target_file"
+
+    case "$original_file" in
+        (*/id_*.pub|*/known_hosts)
```
IMHO easier to match on the key files, e.g. with 
```bash
for key in $ROOTFS_DIR/root/.ssh/id* ; do
  [[ "$key" == *.pub ]] && continue
  encrypt or delete key file
done
```

But that belongs into the `build` stage actually.

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -3,164 +3,97 @@
+    case "$original_file" in
+        (*/id_*.pub|*/known_hosts)
+            # Public key files or 'knownhosts' contain no secrets and can be copied as-is
+            Log "Copying $original_file"
```
I wouldn't log this. It is only interesting for you right now while you develop it. I would log the encryption and the deletion because it is something beyond just copying. In verbose mode the copied files show up in the log in any case.

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -3,164 +3,97 @@
-                    Log "Copying protected private key $original_file retaining its original passphrase"
-                    LogUserOutput "NOTE: private key $original_file was already passphrase-protected and was copied retaining its original passphrase"
+                    LogPrint "TIP: To have the unprotected SSH private key $original_file copied to the rescue system, see SSH_PRIVATE_KEYS_RECOVER_PASSPHRASE in default.conf"
+                    rm "$target_file"
```
So your simplification is to remove the unprotected keys? Are you sure that we can remove this feature (copy unprotected keys to the rescue system) just like this? Maybe users rely on that and use other measures to secure the access (like IP-based restrictions, SSH force commands or even enabling the access only for the time of the recovery).

IMHO this is too harsh. Please add the suggested variable that permits unprotected SSH keys.

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -3,164 +3,97 @@
-
-fi
+            else
+                Log "Copying protected private key $original_file retaining its original passphrase"
```
IMHO useless log entry, it just does what everybody expects it to do.

schlomo commented on usr/share/rear/setup/default/005_ssh_agent_start.sh:
```diff
@@ -4,8 +4,14 @@
 # Public License. Refer to the included COPYING for full text of license.
 
-if has_binary ssh-agent; then
+if has_binary ssh-agent && has_binary ssh && strings "$(which ssh)" | grep -iq AddKeysToAgent ; then
```
What for the `strings` call? `grep` will also work on the binary itself.

schlomo commented on usr/share/rear/rescue/default/500_ssh.sh:
```diff
@@ -3,164 +3,97 @@
+
+mkdir -p -m u=rwx,go=- "$target_dir"  # COPY_AS_IS will take effect later, so create the target .ssh directory first
+
+for original_file in /root/.ssh/id_* /root/.ssh/known_host[s]; do
```
Looking at that I think of another scenario: Users can specify custom SSH keys in their SSH configuration (Identity option). If the key is in the `.ssh` directory that should "just work" and if it is somewhere else then currently users typically add their key file to `COPY_AS_IS`.

You could go ahead and also inspect the SSH config for `Identity` lines and also look at those keys. Moving all of that into the `build` stage will make that possible as then the `COPY_AS_IS` will already have happened.

jsmeix commented:
Via https://github.com/rear/rear/pull/1505 plus the fix in
https://github.com/rear/rear/commit/0f5220ec76dd68a061b55d30cccde5f8aa8505bb
the UserInput function supports now the 'read' options 
'-r' for raw input and 
'-s' to not echo input from a terminal on that terminal 
plus a better description how to use the confidential UserInput mode
in particular an example how to let the user enter a password
that is (hopefully - unless I forgot something) now really quiet
<pre>
{ password="$( UserInput -I PASSWORD -C -r -s -p 'Enter the pasword' )" ; } 2>/dev/null
</pre>
where also in debugscripts mode nothing appears in the log,
at least as it seems for me.
I (mis)use for such tests init/default/030_update_recovery_system.sh
because that is an 'init' script that is run in any case:
<pre>
e205:~/rear.master # head usr/share/rear/init/default/030_update_recovery_system.sh

{ password="$( UserInput -I PASSWORD -C -r -s -p 'Enter the pasword' )" ; } 2>/dev/null
echo -En "$password" | od -a 1>&7

UserInput -I WAIT -t 0

Error "test exit"

# Updating the currently running system.

e205:~/rear.master # export USER_INPUT_PASSWORD='my \t password'

e205:~/rear.master # usr/sbin/rear -d -D mkrescue
Relax-and-Recover 2.2 / Git
Using log file: /root/rear.master/var/log/rear/rear-e205.log
UserInput -I PASSWORD needed in /root/rear.master/usr/share/rear/init/default/030_update_recovery_system.sh line 2
Enter the pasword
(timeout 300 seconds)
UserInput: Will use predefined input in USER_INPUT_PASSWORD
Hit any key to interrupt the automated input (timeout 5 seconds)
0000000   m   y  sp   \   t  sp   p   a   s   s   w   o   r   d
0000016
UserInput -I WAIT needed in /root/rear.master/usr/share/rear/init/default/030_update_recovery_system.sh line 5
enter your input

UserInput: Neither real user input nor real default input (both empty or only spaces) results ''
ERROR: test exit
Aborting due to an error, check /root/rear.master/var/log/rear/rear-e205.log for details
Terminated

e205:~/rear.master # less /root/rear.master/var/log/rear/rear-e205.log
...
+ source /root/rear.master/usr/share/rear/init/default/030_update_recovery_system.sh
++ od -a
++ echo -En 'my \t password'
++ UserInput -I WAIT -t 0
...
</pre>

OliverO2 commented:
@schlomo

Sorry, it seems we have reached a point where I am not going to pursue this PR any further. Let me just clarify:

1. We seem to have completely different philosophies regarding how to deal with security vulnerabilities and how to avoid them:
     - While ReaR normally cherry-picks every bit that should go into the recovery system, you now argue that the potentially vulnerable `/root/.ssh` directory contents should go there completely by default.
     - I think it's not safe to assume that administrators,  when changing stuff in `/root/.ssh`, are fully aware of the consequences in conjunction with ReaR or that they even think about this once ReaR was set up.
     - I think it's not safe to assume that administrators have other protection measures in place. Secrets allowing complete access to sensitive information will end up on unencrypted USB sticks ready for anyone to pick up - an attacker's paradise.
     - So I argue that secrets and other potentially vulnerable configuration files should not be transferred by default. Administrators should decide individually after assessing their specific situation (and probably after thorough consultation with security experts).
     - I simply don't have the time and the necessary in-depth qualification to assess the security implications of every conceivable ssh configuration option, so I'd rather not seduce anyone by providing risky options. I like to play safe and stick to a reasonable minimum with (hopefully) well known consequences.

1. In my view, we are losing focus and wasting time. Examples (all in `500_ssh.sh` unless otherwise stated):
    - I had changed the introductory `if` into `has_binary ssh || has_binary sshd || return` on your request. This was unrelated to the PR's objective, changed indentation everywhere, and the situation got worse after merging a conflicting commit into master.
    - Issues are being discussed which are unrelated to the code actually changed by this PR:
        - `CLONE_USERS`
        - splitting it up into prep and build stage (the file contained code writing to `$ROOTFS_DIR` before)
    - Issues are being discussed which are of too little value to spend time on *now*, for example:
        - When there is `ISO_RECOVER_MODE`, `PXE_RECOVER_MODE` and `LANG_RECOVER`, `SSH_PRIVATE_KEYS_RECOVER_PASSPHRASE` (named after the phase where it is used) should be just OK.
        - The `strings` call in `005_ssh_agent_start.sh` might not be strictly required, but makes sense when you don't want to research `grep`'s implementation details (it's meant to be a text-line-based tool) and the extra call doesn't hurt at all.
        - Using `cat` instead of `$(< file)`, escpecially when it's meant to deliver an empty string if the file doesn't exist, just works. It was just an example in a comment, after all. And opinions on whether `$(< file)` is actually immediately understood by non-bash-experts looking at the configuration seems questionable.
   - I don't say that these things couldn't be discussed but extensive discussions of questionable value are not what I'd call contributor-friendly. I just value my time more than what I think we're getting out of it. And while I'm happy to improve some other details on-the-fly, I did not intend to do a complete overhaul of `500_ssh.sh`.

So if you have very specific ideas of what the code should look like, that's absolutely fine. I'd say just take my ideas and research from this PR and implement it the way you like.

And finally, thanks to you, to the other contributors for creating ReaR, and especially to @jsmeix for being so helpful.

OliverO2 commented:
Topic now tracked in issue #1511.
